spring-cloud-gcp-data-spanner/src/test/java/org/springframework/cloud/gcp/data/spanner/test/AbstractSpannerIntegrationTest.java

```diff
@@ -0,0 +1,180 @@
+import com.google.cloud.spanner.DatabaseAdminClient;
+import com.google.cloud.spanner.DatabaseId;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.BeforeClass;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.config.ConfigurableListableBeanFactory;
+import org.springframework.beans.factory.support.DefaultListableBeanFactory;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.cloud.gcp.data.spanner.core.mapping.SpannerPersistentEntity;
+import org.springframework.cloud.gcp.data.spanner.core.mapping.SpannerPersistentEntityImpl;
+import org.springframework.cloud.gcp.data.spanner.test.domain.Trade;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.ConfigurableApplicationContext;
+import org.springframework.data.util.TypeInformation;
+
+import static org.hamcrest.Matchers.is;
+import static org.junit.Assume.assumeThat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+/**
+ * This class provides the foundation for the integration test framework for Spanner.
+ * Its responsibilities:
+ * <ul>
+ *   <li>initializes the Spring application context</li>
+ *   <li>sets up the database schema</li>
+ *   <li>manages table suffix generation so that parallel test cases don't collide</li>
+ * </ul>
+ * <p>
+ * Prerequisites for running integration tests:
+ *
+ * For Spanner integration tests, you will need to have an instance predefined, everything
+ * else is generated. The instance by default is "integration-instance", which can be
+ * overriden in <code>src/test/resources/application-test.properties</code>, the property:
+ * <code>test.integration.spanner.instance</code>
+ * </p>
+ * <p>
+ * Within the <code>integration-instance</code> instance, the tests rely on a single database for
+ * tests, <code>integration-db</code>. This is automatically created, if doesn't exist.
+ * The tables are generated to have a unique postfix, which is updated on the entity
+ * annotations as well dynamically to avoid collisions of multiple parallel tests running
+ * against the same instance.
+ * </p>
+ *
+ * @author Balint Pato
+ */
+@SpringBootTest(classes = { IntegrationTestConfiguration.class })
+public abstract class AbstractSpannerIntegrationTest {
+
+	private static final String TABLE_NAME_SUFFIX_BEAN_NAME = "tableNameSuffix";
+
+	@Autowired
+	protected DatabaseAdminClient databaseAdminClient;
+
+	@Autowired
+	protected DatabaseId databaseId;
+
+	@Autowired
+	protected ApplicationContext applicationContext;
+
+	protected String tableNameSuffix;
+
+	private boolean setupFailed;
+
+	@BeforeClass
+	public static void checkToRun() {
+		assumeThat("Spanner integration tests are disabled. Please use '-Dit.spanner=true' "
+				+ "to enable them. ",
+				System.getProperty("it.spanner"),
+				is("true"));
+	}
+
+	@Before
+	public void setup() {
+		try {
+			createDatabaseWithSchema();
```

Comment:
The problem with `@Before` that it is going to be called really *before* each test method.
With this `createDatabaseWithSchema()` I'm afraid we are going to spend too much time for each test method...

Can't we really distinguish database creating/deleting from the real application context for testing?
I mean create DB in the `@BeforeClass` and remove it in the `@AfterClass`.

In the end you really can use `static` properties as references in `@Bean` methods.

My point is performance and unnecessary overhead before/after each test method.

WDYT?

Reply:
1.) Performance: Very true, this is an overhead, it is 30 seconds. I think we can make that optimization if it becomes an issue to share a created database between the test methods of a test class. For now this is sufficient for the single test methods per each test class. I wouldn't share the same schema across test classes, that limits parallelisation possibilities across classes. I would be even careful with sharing within class too if we want to run test methods in parallel potentially. 
2.) Implementation of sharing within a class: 
a.) `@BeforeClass` requires static method: I am unclear how I could cleanly use beans from a static method though...I need the `databaseAdminClient` bean and currently I share the same `applicationContext` between the setup method and the test methods so that the `tableNameSuffix` is the same. 
b.) Maybe just use member variables to control the flow. Could definitely be another PR at some point if we see the need for more integration tests.

Reply:
OK. We can reconsider this whenever the problem will be raised.
As far as I see these tests are not going to be run all the time. Only in case of that enabled property for integration tests.